File: BaseSDK/Windows/Samples/Base/DS86/ColorResolutionChange/ColorResolutionChange.cpp
#include <thread>
#include <iostream>
#include "Scepter_api.h"
#define frameSpace 10
using namespace std;

int main()
{
	cout << "---ColorResolutionChange---"<< endl;

	//about dev
	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;
	
	//about frame
	
	ScFrameReady FrameReady = {0};
	ScFrame ColorFrame = { 0 };

	//SDK Initialize
	status = scInitialize();
	if (status != ScStatus::SC_OK)
	{
		cout << "ScInitialize failed status:" <<status << endl;
		system("pause");
		return -1;
	}

	//1.Search and notice the count of devices.
	//2.get infomation of the devices. 
	//3.open devices accroding to the info.
	status = scGetDeviceCount(&deviceCount, 3000);
	if (status != ScStatus::SC_OK)
	{
		cout << "ScGetDeviceCount failed! make sure pointer valid or called scInitialize()" << endl;
		system("pause");
		return -1;
	}
	cout << "Get device count: " << deviceCount << endl;
	if (0 == deviceCount)
	{
		cout << "scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples."<< endl;
		system("pause");
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << "connect status: " << pDeviceListInfo[0].status << endl;
			cout << "The device state does not support connection."<< endl;
			return -1;
		}
	}
	else
	{
		cout << "GetDeviceListInfo failed status:" << status << endl;
		return -1;
	}

	cout << "serialNumber:" << pDeviceListInfo[0].serialNumber << endl
	<< "ip:" << pDeviceListInfo[0].ip << endl
	<< "connectStatus:" << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "OpenDevice failed status:" <<status << endl;
		return false;
	}

    cout << "open device successful,status :" << status << endl;

	//switch ColorResolution
	int resolution_w = 640;
	int resolution_h = 480;
	//1.640_480
	status = scSetColorResolution(deviceHandle, resolution_w, resolution_h);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetColorResolution failed status:" <<status<< endl;
		return -1;
	}
	else
	{
		cout << "set to 640_480" << endl;
	}

	//Starts capturing the image stream
	status = scStartStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStartStream failed status:" << status << endl;
		return -1;
	}

	//scGetFrameReady may return SC_GET_FRAME_READY_TIME_OUT.
	//Because it takes a little of time for scStartStream to upload image data.
	for (int i = 0; i < frameSpace; i++)
	{
		status = scGetFrameReady(deviceHandle, 1200, &FrameReady);
		if (status != ScStatus::SC_OK)
		{
			cout << "scGetFrameReady failed status:" <<status<< endl;
			continue;
		}

		//cout resolution
		if (1 == FrameReady.color)
		{
			status = scGetFrame(deviceHandle, SC_COLOR_FRAME, &ColorFrame);
			if (status == ScStatus::SC_OK && ColorFrame.pFrameData != NULL
			&& ColorFrame.width==640 )
			{
				cout << "get Frame successful,status:" << status << "  "
					 << "resolution: "<< ColorFrame.width  <<"x"<<ColorFrame.height<< endl;	
			}
		}

	}

	resolution_w = 1600;
	resolution_h = 1200;
	//2.1600_1200
	status = scSetColorResolution(deviceHandle, resolution_w, resolution_h);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetColorResolution failed status:" <<status<< endl;
		return -1;
	}
	else
	{
		cout << "set to 1600_1200" << endl;
	}

	//scGetFrameReady may return SC_GET_FRAME_READY_TIME_OUT.
	//Because it takes a little of time for scSetColorResolution to upload image data.
	for (int i = 0; i < frameSpace; i++)
	{
		status = scGetFrameReady(deviceHandle, 1200, &FrameReady);
		if (status != ScStatus::SC_OK)
		{
			cout << "scGetFrameReady failed status:" <<status<< endl;
			continue;
		}

		//cout resolution
		if (1 == FrameReady.color)
		{
			status = scGetFrame(deviceHandle, SC_COLOR_FRAME, &ColorFrame);
			if (status == ScStatus::SC_OK && ColorFrame.pFrameData != NULL
						&& ColorFrame.width==1600 )

			{
				cout << "get Frame successful,status:" << status << "  "
					<< "resolution: "<< ColorFrame.width  <<"x"<<ColorFrame.height<< endl;	
			}
		}

	}
	//Stop capturing the image stream
	status = scStopStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStopStream failed status:" <<status<< endl;
		return -1;
	}

	//1.close device
	//2.SDK shutdown
	status = scCloseDevice(&deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scCloseDevice failed status:" <<status<< endl;
		return -1;
	}
	status = scShutdown();
	if (status != ScStatus::SC_OK)
	{
		cout << "scShutdown failed status:" <<status<< endl;
		return -1;
	}
	cout<< "---end---"<< endl;

	return 0;
}

File: MultilanguageSDK/Python/Samples/DS86/ColorResolutionChange/ColorResolutionChange.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time

frameSpace = 10
camera = ScepterTofCam()


camera_count = camera.scGetDeviceCount(3000)
print("Get device count:", camera_count)
if camera_count <= 0:
    print("scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
    else:
        print(' failed:' + ret)  
        exit()  
else: 
    print("there are no camera found")
    exit()

if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
	print("connect status:",device_info.status)  
	print("Call scOpenDeviceBySN with connect status :",ScConnectStatus.SC_CONNECTABLE.value)
	exit()
else:
    print("serialNumber: "+str(device_info.serialNumber))
    print("ip: "+str(device_info.ip))
    print("connectStatus: "+str(device_info.status))

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret != 0:
    print('scOpenDeviceBySN failed: ' + str(ret))
    exit()
	
print("scOpenDeviceBySN,status :"+ str(ret))

ret = camera.scStartStream()
if  ret != 0:
    print("scStartStream failed:"+ str(ret))
    exit()     


ret = camera.scSetColorResolution(640, 480)
time.sleep(2)
if  ret != 0:  
    print("scSetColorResolution failed status:",ret)
else:
    print("set to 640_480") 

for i in range(frameSpace):
    ret, frameready = camera.scGetFrameReady(c_uint16(1200))
    if  ret !=0:
        print("scGetFrameReady failed status:",ret)
        continue       
    
    if  frameready.color:      
        ret,frame = camera.scGetFrame(ScFrameType.SC_COLOR_FRAME)
        if  ret == 0:
            if frame.width == 640 and frame.height == 480:
                print( "scGetFrame status:" , ret, " resolution: ", frame.width  ,"x",frame.height)
            else:
                print("color width ",frame.width,"  height ",frame.height)      
        else:   
            print("color  error:",ret)

ret = camera.scSetColorResolution(1600, 1200)
time.sleep(2)
if  ret != 0:  
    print("scSetColorResolution failed status:",ret)
else:
    print("set to 1600_1200") 

for i in range(frameSpace):
    ret, frameready = camera.scGetFrameReady(c_uint16(1200))
    if  ret !=0:
        print("scGetFrameReady failed status:",ret)
        continue       
    
    if  frameready.color:      
        ret,frame = camera.scGetFrame(ScFrameType.SC_COLOR_FRAME)
        if  ret == 0:
            if frame.width == 1600 and frame.height == 1200:
                print( "scGetFrame status:" , ret, " resolution: ", frame.width  ,"x",frame.height)
            else:
                print("color width ",frame.width,"  height ",frame.height)      
        else:   
            print("color  error:",ret)


ret = camera.scStopStream()
if  ret == 0:
    print("stopstream success")
else:
    print("stopstream failed",ret)

ret = camera.scCloseDevice()     
if  ret == 0:
    print("scCloseDevice successful")
else:
    print('scCloseDevice failed: ' + str(ret)) 
           
